waku/v2/protocol/filter/waku_filter.go

```diff
@@ -0,0 +1,278 @@
+		} else {
+			// TODO wf.subscribers.unsubscribeFilters(filterRPCRequest.Request, conn.peerInfo.peerId)
+		}
+	} else if filterRPCRequest.Push != nil {
+		// We're on a light node.
+		// This is a message push coming from a full node.
+
+		log.Info("Light node, received a message push ", *filterRPCRequest.Push)
+		wf.pushHandler(filterRPCRequest.RequestId, *filterRPCRequest.Push)
+	}
+
+}
+
+func NewWakuFilter(ctx context.Context, host host.Host, handler MessagePushHandler) *WakuFilter {
+	wf := new(WakuFilter)
+	wf.ctx = ctx
+	wf.MsgC = make(chan *protocol.Envelope)
+	wf.h = host
+	wf.pushHandler = handler
+
+	wf.h.SetStreamHandler(WakuFilterProtocolId, wf.onRequest)
+	go wf.FilterListener()
+
+	return wf
+}
+
+func (wf *WakuFilter) FilterListener() {
+
+	// This function is invoked for each message received
+	// on the full node in context of Waku2-Filter
+	handle := func(topic string, msg *pb.WakuMessage) error { // async
+		// trace "handle WakuFilter subscription", topic=topic, msg=msg
+
+		// Each subscriber is a light node that earlier on invoked
+		// a FilterRequest on this node
+		for _, subscriber := range wf.subscribers {
+			// TODO
+			// if subscriber.filter.pubSubTopic != "" && subscriber.filter.pubSubTopic != topic {
+			// 	log.Info("Subscriber's filter pubsubTopic does not match message topic", subscriber.filter.pubSubTopic, topic)
+			// 	continue
+			// }
+
+			for _, filter := range subscriber.filter.ContentFilters {
+				if msg.ContentTopic == filter.ContentTopic {
+					log.Info("Found matching contentTopic", filter, msg)
+					msgArr := []*pb.WakuMessage{msg}
+					// Do a message push to light node
+					pushRPC := &pb.FilterRPC{RequestId: subscriber.requestId, Push: &pb.MessagePush{Messages: msgArr}}
+					log.Info("Pushing a message to light node: ", pushRPC)
+
+					conn, err := wf.h.NewStream(wf.ctx, peer.ID(subscriber.peer), WakuFilterProtocolId)
+
+					if err != nil {
+						// @TODO more sophisticated error handling here
+						log.Error("Failed to open peer stream")
+						//waku_filter_errors.inc(labelValues = [dialFailure])
+						return err
+					}
+					writer := protoio.NewDelimitedWriter(conn)
+					err = writer.WriteMsg(pushRPC)
+					if err != nil {
+						log.Error("failed to push messages to remote peer")
+						return nil
+					}
+
+				}
+			}
+		}
+
+		return nil
+	}
+
+	for m := range wf.MsgC {
+		handle("", m.Message())
+	}
+
+}
+
+// TODO Remove code duplication
+func (wf *WakuFilter) AddPeer(p peer.ID, addrs []ma.Multiaddr) error {
+	for _, addr := range addrs {
+		wf.h.Peerstore().AddAddr(p, addr, peerstore.PermanentAddrTTL)
+	}
+	err := wf.h.Peerstore().AddProtocols(p, string(WakuFilterProtocolId))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Having a FilterRequest struct,
+// select a peer with filter support, dial it,
+// and submit FilterRequest wrapped in FilterRPC
+func (wf *WakuFilter) Subscribe(request pb.FilterRequest) (string, error) { //.async, gcsafe.} {
```

Comment:
Instead of reusing the context stored in `wf`, consider receiving the context as a parameter of the `Subscribe` function, that way it should be possible to pass a custom context that can timeout or can be canceled

Reply:
Yes, thanks, done.